<a href="https://colab.research.google.com/github/christophermalone/DSCI325/blob/main/Module3_Part3_R.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Module 3 | Part 3 | R : MUTATE()

This notebook will discuss the use of the MUTATE() data verb in R.

<table width='100%' ><tr><td bgcolor='green'></td></tr></table>

### Example 3.3.1R
For this notebook, consider the Nursing Home Provider dataset.  The datasets is published by the Centers for Medicare and Medicade Services.

<table>
<tr>
<td colspan="2" align='center'>Fields in Dataset</td>
</tr>
<tr>
<td valign='top' align='left'>
1: FPN (Federal Provider Number)<br>
2: ProviderName<br>
3: ProviderAddress<br>
4: ProviderCounty<br>
5: OwnershipType<br><br>
6: NumberBeds<br>
7: AverageResidentsperDay<br><br>
8: ProviderInHospital<br>
9: BusinessName<br>
10: ContinuingCareRetirementCommunity<br><br>
11: HealthInspection_MoreThan2YearsAgo<br>
12: Overall_Rating<br>
13: HealthInspection_Rating<br>
14: QM_Rating<br>
15: LongStayQM_Rating<br>
16: ShortStayQM_Rating<br>
17: Staffing_Rating<br>
18: RNStaffing_Rating<br>
</td>
<td valign='top' align='left'>
19: NurseAideStaffing_Hours<br>
20: LPNStaffing_Hours<br>
21: RNStaffing_Hours<br>
22: LicensedStaffing_Hours<br>
23: TotalNurseStaffing_Hours<br>
24: PhysicalTherapistStaffing_Hours<br><br>
25: RatingCycle1_Date<br>
26: RatingCycle1_HealthDeficiencyScore<br>
27: RatingCycle2_Date<br>
28: RatingCycle2_HealthDeficiencyScore<br>
29: RatingCycle3_Date<br>
30: RatingCycle3_HealthDeficiencyScore<br><br>
31: NumberofFines<br>
32: TotalAmountFines<br>
</td>
</tr>
</table>

 
<br>Original Data Source: https://data.cms.gov/provider-data/search?theme=Nursing%20homes%20including%20rehab%20services

<table width='100%' ><tr><td bgcolor='green'></td></tr></table>

Load the tidyverse() package.

In [1]:
library(tidyverse)

Warning message in system("timedatectl", intern = TRUE):
“running command 'timedatectl' had status 1”
── Attaching packages ─────────────────────────────────────── tidyverse 1.3.0 ──

✔ ggplot2 3.3.3     ✔ purrr   0.3.4
✔ tibble  3.1.0     ✔ dplyr   1.0.4
✔ tidyr   1.1.3     ✔ stringr 1.4.0
✔ readr   1.4.0     ✔ forcats 0.5.1

── Conflicts ────────────────────────────────────────── tidyverse_conflicts() ──
✖ dplyr::filter() masks stats::filter()
✖ dplyr::lag()    masks stats::lag()



Reading in the ProviderInfo.csv dataset using read_csv().

In [3]:
NursingHomes <- read_csv("/content/sample_data/ProviderInfo.csv")


── Column specification ────────────────────────────────────────────────────────
cols(
  .default = col_double(),
  FPN = col_character(),
  ProviderName = col_character(),
  ProviderAddress = col_character(),
  ProviderCounty = col_character(),
  OwnershipType = col_character(),
  ProviderInHospital = col_character(),
  BusinessName = col_character(),
  ContinuingCareRetirementCommunity = col_character(),
  HealthInspection_MoreThan2YearsAgo = col_character(),
  RatingCycle1_Date = col_character(),
  RatingCycle1_HealthDeficiencyScore = col_character(),
  RatingCycle2_Date = col_character(),
  RatingCycle2_HealthDeficiencyScore = col_character(),
  RatingCycle3_Date = col_character(),
  RatingCycle3_HealthDeficiencyScore = col_character()
)
ℹ Use `spec()` for the full column specifications.




In [12]:
#Getting the dimension
cat("The number of rows and columns in this dataframe is: \n")
dim(NursingHomes)
cat("\n")


#Getting a list of all the field names
cat("Field names in the Nursing Homes Dataset: \n")
colnames(NursingHomes)

#Seeing first few rows
head(NursingHomes)


The number of rows and columns in this dataframe is: 


[1] 15331    32


Field names in the Nursing Homes Dataset: 


[1] "FPN"                                "ProviderName"                      
 [3] "ProviderAddress"                    "ProviderCounty"                    
 [5] "OwnershipType"                      "NumberBeds"                        
 [7] "AverageResidentsperDay"             "ProviderInHospital"                
 [9] "BusinessName"                       "ContinuingCareRetirementCommunity" 
[11] "HealthInspection_MoreThan2YearsAgo" "Overall_Rating"                    
[13] "HealthInspection_Rating"            "QM_Rating"                         
[15] "LongStayQM_Rating"                  "ShortStayQM_Rating"                
[17] "Staffing_Rating"                    "RNStaffing_Rating"                 
[19] "NurseAideStaffing_Hours"            "LPNStaffing_Hours"                 
[21] "RNStaffing_Hours"                   "LicensedStaffing_Hours"            
[23] "TotalNurseStaffing_Hours"           "PhysicalTherapistStaffing_Hours"   
[25] "RatingCycle1_Date"                  "RatingCycle1_HealthDeficiencyScore"
[27] "RatingCycle2_Date"                  "RatingCycle2_HealthDeficiencyScore"
[29] "RatingCycle3_Date"                  "RatingCycle3_HealthDeficiencyScore"
[31] "NumberofFines"                      "TotalAmountFines"

FPN   ProviderName                               
1 15009 BURNS NURSING HOME, INC.                   
2 15010 COOSA VALLEY HEALTHCARE CENTER             
3 15012 HIGHLANDS HEALTH AND REHAB                 
4 15014 EASTVIEW REHABILITATION & HEALTHCARE CENTER
5 15015 PLANTATION MANOR NURSING HOME              
6 15016 ATHENS HEALTH AND REHABILITATION LLC       
  ProviderAddress                               ProviderCounty      
1 701 MONROE STREET NW,RUSSELLVILLE,AL,35653    Franklin County, AL 
2 260 WEST WALNUT STREET,SYLACAUGA,AL,35150     Talladega County, AL
3 380 WOODS COVE ROAD,SCOTTSBORO,AL,35768       Jackson County, AL  
4 7755 FOURTH AVENUE SOUTH,BIRMINGHAM,AL,35206  Jefferson County, AL
5 6450 OLD TUSCALOOSA HIGHWAY,MC CALLA,AL,35111 Jefferson County, AL
6 611 WEST MARKET STREET,ATHENS,AL,35611        Limestone County, AL
  OwnershipType            NumberBeds AverageResidentsperDay ProviderInHospital
1 For profit - Corporation  57         51.8                  N                 
2 For profit - Corporation  85         72.1                  N                 
3 Government - County       50         41.3                  Y                 
4 For profit - Individual   92         71.8                  N                 
5 For profit - Individual  103         83.8                  N                 
6 For profit - Corporation 149        125.2                  N                 
  BusinessName                         ContinuingCareRetirementCommunity ⋯
1 BURNS NURSING HOME, INC.             N                                 ⋯
2 COOSA VALLEY HEALTHCARE CENTER LLC   N                                 ⋯
3 JACKSON COUNTY HEALTH CARE AUTHORITY N                                 ⋯
4 BALL HEALTHCARE EASTVIEW INC         N                                 ⋯
5 C & G HEALTHCARE SERVICES, INC.      N                                 ⋯
6 ATHENS HEALTH AND REHABILITATION LLC N                                 ⋯
  TotalNurseStaffing_Hours PhysicalTherapistStaffing_Hours RatingCycle1_Date
1 4.31259                  0.00000                         8/21/2019        
2 4.84021                  0.01271                         6/13/2019        
3 4.76645                  0.05596                         6/6/2019         
4 3.81920                  0.03246                         2/20/2020        
5 3.39949                  0.01248                         3/5/2020         
6 3.24110                  0.03716                         8/29/2019        
  RatingCycle1_HealthDeficiencyScore RatingCycle2_Date
1 8                                  8/1/2018         
2 4                                  6/7/2018         
3 20                                 5/3/2018         
4 4                                  1/24/2019        
5 20                                 5/2/2019         
6 20                                 7/19/2018        
  RatingCycle2_HealthDeficiencyScore RatingCycle3_Date
1 4                                  6/22/2017        
2 32                                 4/6/2017         
3 40                                 3/16/2017        
4 20                                 1/25/2018        
5 4                                  4/19/2018        
6 28                                 6/1/2017         
  RatingCycle3_HealthDeficiencyScore NumberofFines TotalAmountFines
1 0                                  0                 0           
2 36                                 1               650           
3 44                                 0                 0           
4 24                                 0                 0           
5 28                                 2             29611           
6 20                                 0                 0

# Create New Field - Numeric

Nursing homes are most profitable when they have lots of residents, i.e. are near capacity.  Consider the following new field to be calculated for this data.

$$Capacity = \frac{Average \space Residents \space per \space Day}{Number \space of \space Beds}$$

The <strong>mutate()</strong> function in the <strong>dplyr</strong> package will can be used to complete this task.  The mutate() function has the following syntax.

In [28]:
(
  NursingHomes
  %>% mutate(Capacity = AverageResidentsperDay / NumberBeds)
  %>% head()
) 

FPN   ProviderName                               
1 15009 BURNS NURSING HOME, INC.                   
2 15010 COOSA VALLEY HEALTHCARE CENTER             
3 15012 HIGHLANDS HEALTH AND REHAB                 
4 15014 EASTVIEW REHABILITATION & HEALTHCARE CENTER
5 15015 PLANTATION MANOR NURSING HOME              
6 15016 ATHENS HEALTH AND REHABILITATION LLC       
  ProviderAddress                               ProviderCounty      
1 701 MONROE STREET NW,RUSSELLVILLE,AL,35653    Franklin County, AL 
2 260 WEST WALNUT STREET,SYLACAUGA,AL,35150     Talladega County, AL
3 380 WOODS COVE ROAD,SCOTTSBORO,AL,35768       Jackson County, AL  
4 7755 FOURTH AVENUE SOUTH,BIRMINGHAM,AL,35206  Jefferson County, AL
5 6450 OLD TUSCALOOSA HIGHWAY,MC CALLA,AL,35111 Jefferson County, AL
6 611 WEST MARKET STREET,ATHENS,AL,35611        Limestone County, AL
  OwnershipType            NumberBeds AverageResidentsperDay ProviderInHospital
1 For profit - Corporation  57         51.8                  N                 
2 For profit - Corporation  85         72.1                  N                 
3 Government - County       50         41.3                  Y                 
4 For profit - Individual   92         71.8                  N                 
5 For profit - Individual  103         83.8                  N                 
6 For profit - Corporation 149        125.2                  N                 
  BusinessName                         ContinuingCareRetirementCommunity ⋯
1 BURNS NURSING HOME, INC.             N                                 ⋯
2 COOSA VALLEY HEALTHCARE CENTER LLC   N                                 ⋯
3 JACKSON COUNTY HEALTH CARE AUTHORITY N                                 ⋯
4 BALL HEALTHCARE EASTVIEW INC         N                                 ⋯
5 C & G HEALTHCARE SERVICES, INC.      N                                 ⋯
6 ATHENS HEALTH AND REHABILITATION LLC N                                 ⋯
  PhysicalTherapistStaffing_Hours RatingCycle1_Date
1 0.00000                         8/21/2019        
2 0.01271                         6/13/2019        
3 0.05596                         6/6/2019         
4 0.03246                         2/20/2020        
5 0.01248                         3/5/2020         
6 0.03716                         8/29/2019        
  RatingCycle1_HealthDeficiencyScore RatingCycle2_Date
1 8                                  8/1/2018         
2 4                                  6/7/2018         
3 20                                 5/3/2018         
4 4                                  1/24/2019        
5 20                                 5/2/2019         
6 20                                 7/19/2018        
  RatingCycle2_HealthDeficiencyScore RatingCycle3_Date
1 4                                  6/22/2017        
2 32                                 4/6/2017         
3 40                                 3/16/2017        
4 20                                 1/25/2018        
5 4                                  4/19/2018        
6 28                                 6/1/2017         
  RatingCycle3_HealthDeficiencyScore NumberofFines TotalAmountFines Capacity 
1 0                                  0                 0            0.9087719
2 36                                 1               650            0.8482353
3 44                                 0                 0            0.8260000
4 24                                 0                 0            0.7804348
5 28                                 2             29611            0.8135922
6 20                                 0                 0            0.8402685

<u>Question:</u>  Did this new field automatically get added to the NursingHomes dataframe (or did this just get printed to the screen)?

<u>Answer:</u> Check the dimension of the dataframe.  The existing dataframe had 15331 rows x 32 columns.

In [34]:
#What is the dimension of the NursingHomes dataframe?
dim(NursingHomes)

cat("\n")
cat("Field names: \n")
colnames(NursingHomes)

[1] 15331    32


Field names: 


[1] "FPN"                                "ProviderName"                      
 [3] "ProviderAddress"                    "ProviderCounty"                    
 [5] "OwnershipType"                      "NumberBeds"                        
 [7] "AverageResidentsperDay"             "ProviderInHospital"                
 [9] "BusinessName"                       "ContinuingCareRetirementCommunity" 
[11] "HealthInspection_MoreThan2YearsAgo" "Overall_Rating"                    
[13] "HealthInspection_Rating"            "QM_Rating"                         
[15] "LongStayQM_Rating"                  "ShortStayQM_Rating"                
[17] "Staffing_Rating"                    "RNStaffing_Rating"                 
[19] "NurseAideStaffing_Hours"            "LPNStaffing_Hours"                 
[21] "RNStaffing_Hours"                   "LicensedStaffing_Hours"            
[23] "TotalNurseStaffing_Hours"           "PhysicalTherapistStaffing_Hours"   
[25] "RatingCycle1_Date"                  "RatingCycle1_HealthDeficiencyScore"
[27] "RatingCycle2_Date"                  "RatingCycle2_HealthDeficiencyScore"
[29] "RatingCycle3_Date"                  "RatingCycle3_HealthDeficiencyScore"
[31] "NumberofFines"                      "TotalAmountFines"

In the following snipit of code, the outcome for the sequnce of commands is pushed into the NursingHome dataframe.  This idea is equivalent to "edit-in-place" option in SED.

In [35]:
(
  NursingHomes
  %>% mutate(Capacity = AverageResidentsperDay / NumberBeds)
  #%>% head()
) -> NursingHomes 

In [36]:
#What is the dimension of the NursingHomes dataframe now?
dim(NursingHomes)

cat("\n")
cat("Field names: \n")
colnames(NursingHomes)

[1] 15331    33


Field names: 


[1] "FPN"                                "ProviderName"                      
 [3] "ProviderAddress"                    "ProviderCounty"                    
 [5] "OwnershipType"                      "NumberBeds"                        
 [7] "AverageResidentsperDay"             "ProviderInHospital"                
 [9] "BusinessName"                       "ContinuingCareRetirementCommunity" 
[11] "HealthInspection_MoreThan2YearsAgo" "Overall_Rating"                    
[13] "HealthInspection_Rating"            "QM_Rating"                         
[15] "LongStayQM_Rating"                  "ShortStayQM_Rating"                
[17] "Staffing_Rating"                    "RNStaffing_Rating"                 
[19] "NurseAideStaffing_Hours"            "LPNStaffing_Hours"                 
[21] "RNStaffing_Hours"                   "LicensedStaffing_Hours"            
[23] "TotalNurseStaffing_Hours"           "PhysicalTherapistStaffing_Hours"   
[25] "RatingCycle1_Date"                  "RatingCycle1_HealthDeficiencyScore"
[27] "RatingCycle2_Date"                  "RatingCycle2_HealthDeficiencyScore"
[29] "RatingCycle3_Date"                  "RatingCycle3_HealthDeficiencyScore"
[31] "NumberofFines"                      "TotalAmountFines"                  
[33] "Capacity"

<u>Comment</u>: The Capacity quantity is not defined when the denominitor is zero.  Thus, before creating this new field, one should verify that all <i>Number of Beds</i> values are greater than 0. 

A more sound approach, would be to use the following definition of Capacity.


$$Capacity = \Bigg\{ \begin{array} \tiny{\frac{\mbox{Average Residents per Day}}{\mbox{Number of Beds}}} & when & \mbox{Number of Beds} > 0\\
NA &  when & \mbox{Number of Beds} = 0
\end{array}$$



In [41]:
(
  NursingHomes
  %>% mutate(Capacity = ifelse(NumberBeds > 0, AverageResidentsperDay / NumberBeds, NA) )
) -> NursingHomes 


Taking a look at the output produced from the code above.

In [43]:
#What is the dimension of the NursingHomes dataframe now?
dim(NursingHomes)

cat("\n")
cat("Field names: \n")
colnames(NursingHomes)

cat("First few rows: \n")
head(NursingHomes)


[1] 15331    33


Field names: 


[1] "FPN"                                "ProviderName"                      
 [3] "ProviderAddress"                    "ProviderCounty"                    
 [5] "OwnershipType"                      "NumberBeds"                        
 [7] "AverageResidentsperDay"             "ProviderInHospital"                
 [9] "BusinessName"                       "ContinuingCareRetirementCommunity" 
[11] "HealthInspection_MoreThan2YearsAgo" "Overall_Rating"                    
[13] "HealthInspection_Rating"            "QM_Rating"                         
[15] "LongStayQM_Rating"                  "ShortStayQM_Rating"                
[17] "Staffing_Rating"                    "RNStaffing_Rating"                 
[19] "NurseAideStaffing_Hours"            "LPNStaffing_Hours"                 
[21] "RNStaffing_Hours"                   "LicensedStaffing_Hours"            
[23] "TotalNurseStaffing_Hours"           "PhysicalTherapistStaffing_Hours"   
[25] "RatingCycle1_Date"                  "RatingCycle1_HealthDeficiencyScore"
[27] "RatingCycle2_Date"                  "RatingCycle2_HealthDeficiencyScore"
[29] "RatingCycle3_Date"                  "RatingCycle3_HealthDeficiencyScore"
[31] "NumberofFines"                      "TotalAmountFines"                  
[33] "Capacity"

First few rows: 


FPN   ProviderName                               
1 15009 BURNS NURSING HOME, INC.                   
2 15010 COOSA VALLEY HEALTHCARE CENTER             
3 15012 HIGHLANDS HEALTH AND REHAB                 
4 15014 EASTVIEW REHABILITATION & HEALTHCARE CENTER
5 15015 PLANTATION MANOR NURSING HOME              
6 15016 ATHENS HEALTH AND REHABILITATION LLC       
  ProviderAddress                               ProviderCounty      
1 701 MONROE STREET NW,RUSSELLVILLE,AL,35653    Franklin County, AL 
2 260 WEST WALNUT STREET,SYLACAUGA,AL,35150     Talladega County, AL
3 380 WOODS COVE ROAD,SCOTTSBORO,AL,35768       Jackson County, AL  
4 7755 FOURTH AVENUE SOUTH,BIRMINGHAM,AL,35206  Jefferson County, AL
5 6450 OLD TUSCALOOSA HIGHWAY,MC CALLA,AL,35111 Jefferson County, AL
6 611 WEST MARKET STREET,ATHENS,AL,35611        Limestone County, AL
  OwnershipType            NumberBeds AverageResidentsperDay ProviderInHospital
1 For profit - Corporation  57         51.8                  N                 
2 For profit - Corporation  85         72.1                  N                 
3 Government - County       50         41.3                  Y                 
4 For profit - Individual   92         71.8                  N                 
5 For profit - Individual  103         83.8                  N                 
6 For profit - Corporation 149        125.2                  N                 
  BusinessName                         ContinuingCareRetirementCommunity ⋯
1 BURNS NURSING HOME, INC.             N                                 ⋯
2 COOSA VALLEY HEALTHCARE CENTER LLC   N                                 ⋯
3 JACKSON COUNTY HEALTH CARE AUTHORITY N                                 ⋯
4 BALL HEALTHCARE EASTVIEW INC         N                                 ⋯
5 C & G HEALTHCARE SERVICES, INC.      N                                 ⋯
6 ATHENS HEALTH AND REHABILITATION LLC N                                 ⋯
  PhysicalTherapistStaffing_Hours RatingCycle1_Date
1 0.00000                         8/21/2019        
2 0.01271                         6/13/2019        
3 0.05596                         6/6/2019         
4 0.03246                         2/20/2020        
5 0.01248                         3/5/2020         
6 0.03716                         8/29/2019        
  RatingCycle1_HealthDeficiencyScore RatingCycle2_Date
1 8                                  8/1/2018         
2 4                                  6/7/2018         
3 20                                 5/3/2018         
4 4                                  1/24/2019        
5 20                                 5/2/2019         
6 20                                 7/19/2018        
  RatingCycle2_HealthDeficiencyScore RatingCycle3_Date
1 4                                  6/22/2017        
2 32                                 4/6/2017         
3 40                                 3/16/2017        
4 20                                 1/25/2018        
5 4                                  4/19/2018        
6 28                                 6/1/2017         
  RatingCycle3_HealthDeficiencyScore NumberofFines TotalAmountFines Capacity 
1 0                                  0                 0            0.9087719
2 36                                 1               650            0.8482353
3 44                                 0                 0            0.8260000
4 24                                 0                 0            0.7804348
5 28                                 2             29611            0.8135922
6 20                                 0                 0            0.8402685



---

---





**Task #1**:  Consider the following code -- answer the questions that follow. 


In [50]:
(
  NursingHomes
  %>% mutate(Capacity = ifelse(NumberBeds > 0, AverageResidentsperDay / NumberBeds, NA) )
  %>% select(FPN, ProviderName, Capacity)
  %>% arrange(desc(Capacity))
  %>% filter(Capacity > 1)
  %>% dim()
)

[1] 107   3


1.   What is the purpose of Line 4? Discusss

2.   What is the purpose of Line 5? Discuss

3.   What is is the purpose of Line 6? Discuss

4.   What information is gained by running this code?  Discuss



---

---





**Task #2**:  Consider the following code -- answer the questions that follow. 

In [66]:
(
  NursingHomes
  %>% mutate(Capacity = ifelse(NumberBeds > 0, 100*round(AverageResidentsperDay / NumberBeds,3), NA) )
  %>% select(FPN, ProviderName, Capacity)
  %>% arrange(desc(Capacity))
  %>% filter(Capacity <= 100)
  %>% mutate(Capacity = as.character(Capacity))
  %>% mutate(Capacity = paste(Capacity, "%", sep=""))
  %>% head(10)
)


FPN    ProviderName                                      Capacity
1  35287  HAVASU REGIONAL MEDICAL CENTER                    100%    
2  05A192 THE ROYAL HOME                                    100%    
3  16E668 PALO ALTO COUNTY HOSPITAL                         100%    
4  335874 THE STEVEN AND ALEXANDRA COHEN PED L T C PAVILION 100%    
5  495405 SUMMIT SQUARE                                     100%    
6  515182 MILETREE CENTER                                   100%    
7  555450 AMERICAN RIVER CENTER                             100%    
8  105138 HEALTHCARE CENTER OF JACKSONVILLE                 99.8%   
9  675509 PARKLANE WEST HEALTHCARE CENTER                   99.8%   
10 235624 PORTAGEPOINTE                                     99.7%

1.   Change this code to output 20 nursing homes instead of the list provided here.

2.   What would happen if Line 7 were placed before Line 6?  How would this change effect Line 6?

3.   Figure out how to save a csv file of this output. Put this csv file into the '/content/sample_data/' Colab folder.



---

---







Next, calculate the following quantity for each nursing home.

$$ Average \space Fine = \frac{\mbox{Total Amount of Fines}}{\mbox{Number of Fines} }$$

Before, proceeding one should check that this quantity can be computed for all nursing homes. If the Number of Fines is 0, then this quantity is not defined.

In [ ]:
#Getting a summary of all fields in R
summary(NursingHomes)


In [72]:
#Getting the summary for Number of Fines only
summary(NursingHomes$NumberofFines)

   Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
 0.0000  0.0000  0.0000  0.4342  1.0000  9.0000 

Given that many of the nursing homes do not have any fines, the followign should be used for the calcuation of average fine amount.

$$AverageFine = \Bigg\{ \begin{array} \tiny{\frac{\mbox{Total Amount of Fines}}{\mbox{Number of Fines}}} & when & \mbox{Number of Fines} > 0\\
0 &  when & \mbox{Number of Fines} = 0
\end{array}$$

In [74]:
(
  NursingHomes
  %>% mutate(AvgFine = ifelse(NumberofFines > 0, TotalAmountFines / NumberofFines, 0) )
) -> NursingHomes

#Making sure column was added
dim(NursingHomes)
cat("\n")
colnames(NursingHomes)


[1] 15331    34

[1] "FPN"                                "ProviderName"                      
 [3] "ProviderAddress"                    "ProviderCounty"                    
 [5] "OwnershipType"                      "NumberBeds"                        
 [7] "AverageResidentsperDay"             "ProviderInHospital"                
 [9] "BusinessName"                       "ContinuingCareRetirementCommunity" 
[11] "HealthInspection_MoreThan2YearsAgo" "Overall_Rating"                    
[13] "HealthInspection_Rating"            "QM_Rating"                         
[15] "LongStayQM_Rating"                  "ShortStayQM_Rating"                
[17] "Staffing_Rating"                    "RNStaffing_Rating"                 
[19] "NurseAideStaffing_Hours"            "LPNStaffing_Hours"                 
[21] "RNStaffing_Hours"                   "LicensedStaffing_Hours"            
[23] "TotalNurseStaffing_Hours"           "PhysicalTherapistStaffing_Hours"   
[25] "RatingCycle1_Date"                  "RatingCycle1_HealthDeficiencyScore"
[27] "RatingCycle2_Date"                  "RatingCycle2_HealthDeficiencyScore"
[29] "RatingCycle3_Date"                  "RatingCycle3_HealthDeficiencyScore"
[31] "NumberofFines"                      "TotalAmountFines"                  
[33] "Capacity"                           "AvgFine"

Next, consider the average (and median) for all nursing homes.

In [77]:
(
  NursingHomes
  %>% summarize('Average of Fines Overall' = mean(AvgFine), 'Median of Fines Overall' = median(AvgFine))
)

Average of Fines Overall Median of Fines Overall
1 8334.139                 0

Next, compute these quantites for Nursin Home that had a non-zero fine.

In [78]:
(
  NursingHomes
  %>% filter(AvgFine > 0)
  %>% summarize('Average of Fines Overall' = mean(AvgFine), 'Median of Fines Overall' = median(AvgFine))
)

Average of Fines Overall Median of Fines Overall
1 27459.85                 11950

# Create New Field - Character

Next, suppose the desire is to get the average Overall_Rating for nursing homes in MN.  Notice that State is not currently a field.

In [79]:
colnames(NursingHomes)

[1] "FPN"                                "ProviderName"                      
 [3] "ProviderAddress"                    "ProviderCounty"                    
 [5] "OwnershipType"                      "NumberBeds"                        
 [7] "AverageResidentsperDay"             "ProviderInHospital"                
 [9] "BusinessName"                       "ContinuingCareRetirementCommunity" 
[11] "HealthInspection_MoreThan2YearsAgo" "Overall_Rating"                    
[13] "HealthInspection_Rating"            "QM_Rating"                         
[15] "LongStayQM_Rating"                  "ShortStayQM_Rating"                
[17] "Staffing_Rating"                    "RNStaffing_Rating"                 
[19] "NurseAideStaffing_Hours"            "LPNStaffing_Hours"                 
[21] "RNStaffing_Hours"                   "LicensedStaffing_Hours"            
[23] "TotalNurseStaffing_Hours"           "PhysicalTherapistStaffing_Hours"   
[25] "RatingCycle1_Date"                  "RatingCycle1_HealthDeficiencyScore"
[27] "RatingCycle2_Date"                  "RatingCycle2_HealthDeficiencyScore"
[29] "RatingCycle3_Date"                  "RatingCycle3_HealthDeficiencyScore"
[31] "NumberofFines"                      "TotalAmountFines"                  
[33] "Capacity"                           "AvgFine"

Realize that the State can be pulled off from the ProviderAddress or Provider County.  

In [80]:
(
  NursingHomes
  %>% select(ProviderAddress, ProviderCounty)
  %>% head()
)

ProviderAddress                               ProviderCounty      
1 701 MONROE STREET NW,RUSSELLVILLE,AL,35653    Franklin County, AL 
2 260 WEST WALNUT STREET,SYLACAUGA,AL,35150     Talladega County, AL
3 380 WOODS COVE ROAD,SCOTTSBORO,AL,35768       Jackson County, AL  
4 7755 FOURTH AVENUE SOUTH,BIRMINGHAM,AL,35206  Jefferson County, AL
5 6450 OLD TUSCALOOSA HIGHWAY,MC CALLA,AL,35111 Jefferson County, AL
6 611 WEST MARKET STREET,ATHENS,AL,35611        Limestone County, AL

To pull off elements of a string, we will use the <strong>str_sub</strong> function from the <strong>stringr</strong> pakcage. The syntax for this function is

<p align='center'>
<font size='+2'>str_sub(<i>field name</i>, <i>start =</i> , <i>end =</i> )</font>
</p>

Now, negative indicies for <i>start</i> and <i>end</i> can be used to search a string backwards.


In [85]:
(
 NursingHomes
 %>% mutate(State = str_sub(ProviderCounty, -2,-1) )
) -> NursingHomes

colnames(NursingHomes)


[1] "FPN"                                "ProviderName"                      
 [3] "ProviderAddress"                    "ProviderCounty"                    
 [5] "OwnershipType"                      "NumberBeds"                        
 [7] "AverageResidentsperDay"             "ProviderInHospital"                
 [9] "BusinessName"                       "ContinuingCareRetirementCommunity" 
[11] "HealthInspection_MoreThan2YearsAgo" "Overall_Rating"                    
[13] "HealthInspection_Rating"            "QM_Rating"                         
[15] "LongStayQM_Rating"                  "ShortStayQM_Rating"                
[17] "Staffing_Rating"                    "RNStaffing_Rating"                 
[19] "NurseAideStaffing_Hours"            "LPNStaffing_Hours"                 
[21] "RNStaffing_Hours"                   "LicensedStaffing_Hours"            
[23] "TotalNurseStaffing_Hours"           "PhysicalTherapistStaffing_Hours"   
[25] "RatingCycle1_Date"                  "RatingCycle1_HealthDeficiencyScore"
[27] "RatingCycle2_Date"                  "RatingCycle2_HealthDeficiencyScore"
[29] "RatingCycle3_Date"                  "RatingCycle3_HealthDeficiencyScore"
[31] "NumberofFines"                      "TotalAmountFines"                  
[33] "Capacity"                           "AvgFine"                           
[35] "State"

Notice that R adds the newly created columns to the right side of the dataframe.  There exists a <strong>relocate()</strong> function that can be used to rearrange the order of the columns.

In [87]:
(
 NursingHomes
 %>% relocate(State, .after = ProviderAddress )
) -> NursingHomes

colnames(NursingHomes)
cat("\n")
head(NursingHomes)

[1] "FPN"                                "ProviderName"                      
 [3] "ProviderAddress"                    "State"                             
 [5] "ProviderCounty"                     "OwnershipType"                     
 [7] "NumberBeds"                         "AverageResidentsperDay"            
 [9] "ProviderInHospital"                 "BusinessName"                      
[11] "ContinuingCareRetirementCommunity"  "HealthInspection_MoreThan2YearsAgo"
[13] "Overall_Rating"                     "HealthInspection_Rating"           
[15] "QM_Rating"                          "LongStayQM_Rating"                 
[17] "ShortStayQM_Rating"                 "Staffing_Rating"                   
[19] "RNStaffing_Rating"                  "NurseAideStaffing_Hours"           
[21] "LPNStaffing_Hours"                  "RNStaffing_Hours"                  
[23] "LicensedStaffing_Hours"             "TotalNurseStaffing_Hours"          
[25] "PhysicalTherapistStaffing_Hours"    "RatingCycle1_Date"                 
[27] "RatingCycle1_HealthDeficiencyScore" "RatingCycle2_Date"                 
[29] "RatingCycle2_HealthDeficiencyScore" "RatingCycle3_Date"                 
[31] "RatingCycle3_HealthDeficiencyScore" "NumberofFines"                     
[33] "TotalAmountFines"                   "Capacity"                          
[35] "AvgFine"

FPN   ProviderName                               
1 15009 BURNS NURSING HOME, INC.                   
2 15010 COOSA VALLEY HEALTHCARE CENTER             
3 15012 HIGHLANDS HEALTH AND REHAB                 
4 15014 EASTVIEW REHABILITATION & HEALTHCARE CENTER
5 15015 PLANTATION MANOR NURSING HOME              
6 15016 ATHENS HEALTH AND REHABILITATION LLC       
  ProviderAddress                               State ProviderCounty      
1 701 MONROE STREET NW,RUSSELLVILLE,AL,35653    AL    Franklin County, AL 
2 260 WEST WALNUT STREET,SYLACAUGA,AL,35150     AL    Talladega County, AL
3 380 WOODS COVE ROAD,SCOTTSBORO,AL,35768       AL    Jackson County, AL  
4 7755 FOURTH AVENUE SOUTH,BIRMINGHAM,AL,35206  AL    Jefferson County, AL
5 6450 OLD TUSCALOOSA HIGHWAY,MC CALLA,AL,35111 AL    Jefferson County, AL
6 611 WEST MARKET STREET,ATHENS,AL,35611        AL    Limestone County, AL
  OwnershipType            NumberBeds AverageResidentsperDay ProviderInHospital
1 For profit - Corporation  57         51.8                  N                 
2 For profit - Corporation  85         72.1                  N                 
3 Government - County       50         41.3                  Y                 
4 For profit - Individual   92         71.8                  N                 
5 For profit - Individual  103         83.8                  N                 
6 For profit - Corporation 149        125.2                  N                 
  BusinessName                         ⋯ RatingCycle1_Date
1 BURNS NURSING HOME, INC.             ⋯ 8/21/2019        
2 COOSA VALLEY HEALTHCARE CENTER LLC   ⋯ 6/13/2019        
3 JACKSON COUNTY HEALTH CARE AUTHORITY ⋯ 6/6/2019         
4 BALL HEALTHCARE EASTVIEW INC         ⋯ 2/20/2020        
5 C & G HEALTHCARE SERVICES, INC.      ⋯ 3/5/2020         
6 ATHENS HEALTH AND REHABILITATION LLC ⋯ 8/29/2019        
  RatingCycle1_HealthDeficiencyScore RatingCycle2_Date
1 8                                  8/1/2018         
2 4                                  6/7/2018         
3 20                                 5/3/2018         
4 4                                  1/24/2019        
5 20                                 5/2/2019         
6 20                                 7/19/2018        
  RatingCycle2_HealthDeficiencyScore RatingCycle3_Date
1 4                                  6/22/2017        
2 32                                 4/6/2017         
3 40                                 3/16/2017        
4 20                                 1/25/2018        
5 4                                  4/19/2018        
6 28                                 6/1/2017         
  RatingCycle3_HealthDeficiencyScore NumberofFines TotalAmountFines Capacity 
1 0                                  0                 0            0.9087719
2 36                                 1               650            0.8482353
3 44                                 0                 0            0.8260000
4 24                                 0                 0            0.7804348
5 28                                 2             29611            0.8135922
6 20                                 0                 0            0.8402685
  AvgFine
1     0.0
2   650.0
3     0.0
4     0.0
5 14805.5
6     0.0

...  More to follow ....